src/Integration/UserSettingsConfiguration/UserSettingsProvider.cs

```diff
@@ -0,0 +1,248 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2025 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.IO.Abstractions;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.FileMonitor;
+using SonarLint.VisualStudio.Core.Initialization;
+using SonarLint.VisualStudio.Core.SystemAbstractions;
+using SonarLint.VisualStudio.Core.UserRuleSettings;
+using SonarLint.VisualStudio.Infrastructure.VS.Initialization;
+using SonarLint.VisualStudio.Integration.Resources;
+
+namespace SonarLint.VisualStudio.Integration.UserSettingsConfiguration;
+
+[Export(typeof(IUserSettingsProvider))]
+[PartCreationPolicy(CreationPolicy.Shared)]
+internal sealed class UserSettingsProvider : IUserSettingsProvider, IDisposable
+{
+    private const string GeneratedGlobalSettingsFolderName = ".global";
+    private const string SolutionSettingsFolderName = "SolutionSettings";
+    private const string SettingsFileName = "settings.json";
+    private static readonly object Lock = new();
+    private readonly IFileSystem fileSystem;
+    private readonly IActiveSolutionTracker activeSolutionTracker;
+    private readonly ILogger logger;
+    private readonly IAnalysisSettingsSerializer serializer;
+    private readonly ISingleFileMonitorFactory fileMonitorFactory;
+    private ISingleFileMonitor globalSettingsFileMonitor;
+    private ISingleFileMonitor solutionSettingsFileMonitor;
+    private UserSettings userSettings;
+    private string appDataRoot;
+    private bool disposed;
+    private (string settings, string generated) globalFilePaths;
```

Comment:
The name of the tuple is not speaking at all
```suggestion
    private (string fullPath, string folder) globalFilePaths;
```

Reply:
it's not really fullpath and folder, those are independent paths to settings file and to generated folder. 
solutionFilePaths(to)settings, solutionFilePaths(to)generated, etc.

Reply:
`Generated` doesn't tell me anything about what that is. All the folders from the bindings, settings are generated.
You can rename them better, if you can find better names. It was just a suggestion. Other suggestions: `settingsFullPath` and `solutionSettingsFolder`.